### PR TITLE
Cuda persistent softmax

### PR DESCRIPTION
**Pull request by thorjohnsen**
Adds persistent cuda kernels that speed up SoftMax applied over the fast dimension, i.e. torch.nn.Softmax(dim=-1) and torch.nn.LogSoftmax(dim=-1). When the size is <= 1024, this code is 2-10x faster than the current code, speedup is higher for smaller sizes. This code works for half, float and double tensors with 1024 or fewer elements in the fast dimension. Numerical accuracy is on par with the current code, i.e. relative error is ~1e-8 for float tensors and ~1e-17 for double tensors. Relative error was computed against the CPU code.

The attached image shows kernel time in us for torch.nn.Softmax(dim=-1) applied to a half precision tensor of shape [16384,n], n is plotted along the horizontal axis. Similar uplifts can be seen for the backward pass and for LogSoftmax.

image

### CODE REVIEW

**Comment by soumith** [comment]
this is really cool!

**Comment by ngimel** [comment]
Test failures are real.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+#include <stdint.h>
+#include <cuda_fp16.h>
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Warp Softmax forward
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// WARP_BATCH number of batches.
+// WARP_ITERATOINS The number of iterations required for one warp to iterate over all data.
+// WARP_SIZE number of elements working on a single batch, has to be a power of two.
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: number of threads working on a single sample?

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+            max_value[i] = (max_value[i] > elements[i][it]) ? max_value[i] : elements[i][it];
+        }
+    }
+
+    // reduction max_value
+    #pragma unroll
+    for (int offset = WARP_SIZE / 2; offset > 0; offset /= 2) {
+        acc_t val[WARP_BATCH];
+        #pragma unroll
+        for (int i = 0;i < WARP_BATCH;++i) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: Do 2 loops here provide perf benefits, or can they be fused like?
```
val = WARP_SHFL_XOR(..)
max_value[i] = <ternary operator>
```
If 2 loops are indeed necessary, comment might be in order.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+            } else {
+              elements[i][it] = std::exp(elements[i][it] - max_value[i]);
+              sum[i] += elements[i][it];
+            }
+        }
+    }
+
+    // reduction sum
+    #pragma unroll
+    for (int offset = WARP_SIZE / 2; offset > 0; offset /= 2) {
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: I think it would make sense to have a __ device __ function for warp reduce, as it is used a few times. It could also handle max reduction.

**Comment by thorjohnsen** [comment]
Yeah. They are intermittent. I ran one of the failed tests (test_softmax_dtype) two times in a row. It failed the first time and passed the second time. I'm looking into it.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+    // determine size of a warp
+    const int next_power_of_two = 1 << log2_elements;
+    warp_size = (next_power_of_two < 32) ? next_power_of_two : 32;
+
+    // determine how many batches a warp should process.
+    batches_per_warp = (next_power_of_two <= 128) ? 2 : 1;
+
+    switch (log2_elements) {
+    case 0: // 1
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 2,1,1, is_log_softmax>;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: It looks to me like all the template arguments can be constexprs in the kernel, and the only real template argument is `next_power_of_2`. E.g.
```
WARP_SIZE = (np2 <  = 32) ? np2 : 32;
WARP_ITERATIONS = np2/WARP_SIZE;
WARP_BATCH = yet another ternary operator
```
This will make dispatch for forward and backward cleaner.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+        kernel = &softmax_warp_forward<input_t, output_t, acc_t, 1,32,32, is_log_softmax>;
+        break;
+    default:
+        return false;
+    }
+    return true;
+}
+
+template<typename input_t, typename output_t, typename acc_t, bool is_log_softmax>
+bool dispatch_softmax(output_t *dst, const input_t *src, int softmax_elements, int softmax_elements_stride, int batch_count)
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: return value of dispatch_softmax is never checked, so you might as well make it void.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+        // compute launch size
+        int batches_per_block = warps_per_block * batches_per_warp;
+        int blocks = (batch_count + batches_per_block - 1) / batches_per_block;
+        dim3 threads(warp_size, warps_per_block, 1);
+
+        // launch
+        kernel<<<blocks, threads>>>(dst, src, batch_count, softmax_elements_stride, softmax_elements);
+        return true;
+    }
+    return false;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: might make sense to do TORCH_INTERNAL_ASSERT on softmax_elements <=1024.

**Comment by thorjohnsen** [comment]
I'm pretty sure I introduced this bug when I did the code cleanup, the input arrays were no longer being properly initialized. I fixed it, hopefully the tests will pass now.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+
+template<typename input_t, typename output_t, typename acc_t, bool is_log_softmax>
+bool dispatch_softmax(output_t *dst, const input_t *src, int softmax_elements, int softmax_elements_stride, int batch_count)
+{
+    if (softmax_elements == 0) {
+        return true;
+    } else if (softmax_elements <= 1024) {
+        // compute function index. there's a function for each power of two size up to 1024.
+        int log2_elements = 0;
+        while ((1 << log2_elements) < softmax_elements) ++log2_elements;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: @ezyang, @mcarilli the functions for next power of 2 are often needed - does it make sense to make those helper functions generally available?

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+    acc_t sum[WARP_BATCH] = {0};
+    #pragma unroll
+    for (int it = 0;it < WARP_ITERATIONS;++it) {
+        for (int i = 0;i < WARP_BATCH;++i) {
+            sum[i] += grad_reg[i][it];
+        }
+    }
+
+    // reduction sum
+    constexpr uint32_t FULL_MASK = 0xffffffff;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: same comment about FULL_MASK

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+        while ((1 << log2_elements) < softmax_elements) ++log2_elements;
+
+        softmax_backward_func<input_t, output_t> kernel;
+        int warp_size, batches_per_warp;
+        if (!warp_softmax_backward_kernel<input_t, output_t, acc_t, is_log_softmax>(log2_elements, warp_size, batches_per_warp, kernel)) {
+            return false;
+        }
+
+        // use 128 threads per block to maximimize gpu utilization
+        constexpr int threads_per_block = 128;
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: It looks like all these computationas are shared between forward/backward, so may be they can be abstracted away?

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 10]
```diff
@@ -627,19 +649,19 @@ Tensor host_softmax_backward(const Tensor &grad_, const Tensor &output_, int64_t
 }
 
 Tensor log_softmax_cuda(const Tensor &input, const int64_t dim, const bool half_to_float){
-  return host_softmax<LogSoftMaxForwardEpilogue>(input, dim, half_to_float);
+  return host_softmax<LogSoftMaxForwardEpilogue,true>(input, dim, half_to_float);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/SoftMax.cu]: I'm not too happy with 2 template arguments saying the same thing (I'm log! I'm not!), but since I did not come up with a not too ugly way to get rid of it, I'll let it slide.

**Review by ngimel — APPROVED** [review]
Some small changes, overall looks good.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 11]
```diff
@@ -40,6 +40,8 @@ __global__ void softmax_warp_forward(output_t *dst, const input_t *src, int batc
             elements[i][it] = -std::numeric_limits<acc_t>::infinity();
             if (element_index < batch_element_count) {
                 elements[i][it] = src[i*element_count+it*WARP_SIZE];
+            } else {
+                elements[i][it] = -std::numeric_limits<acc_t>::infinity();
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: Why are you setting it to infinity the second time?

**Comment by thorjohnsen** [comment]
One of the softmax tests is still failing on accuracy, but it looks like that is due to a kluge in Pytorch. The test that fails runs softmax on a double precision tensor, but sees float like accuracy (i.e. relative error ~1e-8). The test was compiled with -D__HIP_PLATFORM_HCC__=1. My code uses WARP_SHFL_XOR for intrawarp reductions, if you look at the place where this is defined, there is a comment saying "HIP does not support double" and a specialization of WARP_SHFL_XOR that casts the value to a float before calling __shfl_xor. No test that uses WARP_SHFL_XOR will ever pass the double precision accuracy tests in test_nn.py because of this @iotamudelta.

https://github.com/pytorch/pytorch/blob/31e2d20c5eae46ea2bf1f24ddf96797adda860db/aten/src/THC/THCDeviceUtils.cuh#L63-L68

**Comment by iotamudelta** [comment]
@thorjohnsen A double __shfl_xor(double, int, int) is defined for ROCm. This specialization should be removable.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 12]
```diff
@@ -0,0 +1,385 @@
...
+        for (int it = 0;it < WARP_ITERATIONS;it += 1) {
+            int element_index = local_idx + it * WARP_SIZE;
+            elements[i][it] = -std::numeric_limits<acc_t>::infinity();
+            if (element_index < batch_element_count) {
+                elements[i][it] = src[i*element_count+it*WARP_SIZE];
+            }
+        }
+    }
+
+    constexpr uint32_t  FULL_MASK = 0xffffffff;
```
[the remark this answers is not in the document]
Comment (thorjohnsen) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: I will remove FULL_MASK.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 13]
```diff
@@ -0,0 +1,385 @@
...
+#include <cfloat>
+#include <limits>
+#include <stdint.h>
+#include <cuda_fp16.h>
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+// Warp Softmax forward
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// WARP_BATCH number of batches.
```
[the remark this answers is not in the document]
Comment (thorjohnsen) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: I kind of inherited the comments. I agree they are not very clear, I'll try to improve that.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh (hunk 5 above)
Comment (thorjohnsen) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: I think I prefer to assert on the return value instead. dispatch_softmax will return false if you call it for a case that it can't handle properly.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh (hunk 6 above)
Comment (thorjohnsen) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: That's actually a better idea than asserting on the return value from dispatch_softmax. Will do.

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh (hunk 9 above)
Comment (thorjohnsen) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: Ok.

**Comment by thorjohnsen** [comment]
71 tests passed, 4 failed. After Inspection, I don't think any of the 4 failed tests are related to this PR, but please don't take my word for it. Removing the double specialization of WARP_SHFL_XOR like you suggested fixed the failing Softmax test and a couple of other unrelated unit tests are also passing now, so we might have accidentally fixed a bug @iotamudelta. I will push the new code with your suggested improvements tomorrow morning @ngimel, then I think we should be ready for the merge.

**Comment by ngimel** [comment]
My comments are addressed, thank you.

**Comment by thorjohnsen** [comment]
The 3 failed tests appear unrelated to this PR. One test failed with ImportError torch.onnx.symbolic_helper, the 2nd test failed with ImportError undefined symbol: _ZN2at19NonVariableTypeMode10is_enabledEv and the 3rd test failed because of a java.io.IOException: Backing channel 'JNLP4-connect connection from 209.249.227.2/209.249.227.2:43242' is disconnected. I believe this PR is ready to be merged.
